lib/src/compiler/parse.rs

```diff
@@ -1,14 +1,220 @@
-use crate::{ast::Program, compiler::Compiler};
+use crate::{
+    ast::{Instr, LangValue, Program, StackIdentifier},
+    compiler::Compiler,
+    error::ParseError,
+};
 use failure::Fallible;
+use nom::{
+    branch::alt,
+    bytes::complete::tag,
+    character::complete::{char, digit1, multispace0},
+    combinator::map_res,
+    multi::{many0, many1},
+    sequence::{delimited, preceded, terminated},
+    IResult,
+};
 use std::io::Read;
 
+fn parse_read(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Read")(input)?;
+    Ok((input, Instr::Read))
+}
+
+fn parse_write(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Write")(input)?;
+    Ok((input, Instr::Write))
+}
+
+fn parse_set(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Set")(input)?;
+    let (input, val) = preceded(
+        multispace0,
+        map_res(digit1, |s: &str| s.parse::<LangValue>()),
+    )(input)?;
+    Ok((input, Instr::Set(val)))
+}
+
+fn parse_push(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Push")(input)?;
+    let (input, val) = preceded(
+        multispace0,
+        map_res(digit1, |s: &str| s.parse::<StackIdentifier>()),
+    )(input)?;
+    Ok((input, Instr::Push(val)))
+}
+
+fn parse_pop(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Pop")(input)?;
+    let (input, val) = preceded(
+        multispace0,
+        map_res(digit1, |s: &str| s.parse::<StackIdentifier>()),
+    )(input)?;
+    Ok((input, Instr::Pop(val)))
+}
+
+fn parse_if(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("if")(input)?;
+    let (input, res) = parse_body(input)?;
+    Ok((input, Instr::If(res)))
+}
+
+fn parse_while(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("while")(input)?;
+    let (input, res) = parse_body(input)?;
+    Ok((input, Instr::While(res)))
+}
+
+fn try_each(input: &str) -> IResult<&str, Instr> {
+    let (input, res) = preceded(
+        multispace0,
+        alt((
+            parse_read,
+            parse_write,
+            parse_set,
+            parse_push,
+            parse_pop,
+            parse_if,
+            parse_while,
+        )),
+    )(input)?;
+    Ok((input, res))
+}
+
+fn parse_body(input: &str) -> IResult<&str, Vec<Instr>> {
+    let (input, res) = delimited(
+        preceded(multispace0, char('{')),
+        many0(try_each),
+        preceded(multispace0, char('}')),
+    )(input)?;
+    Ok((input, res))
+}
+
+fn my_parse(input: &str) -> IResult<&str, Vec<Instr>> {
```

Comment:
Can you rename this to be a bit more descriptive?

Reply:
got u fam